runtime/lua/vim/treesitter/query.lua

```diff
@@ -737,87 +790,50 @@ end
 --- Lists the currently available predicates to use in queries.
 ---@return string[] : Supported predicates.
 function M.list_predicates()
-  return vim.tbl_keys(predicate_handlers)
-end
-
-local function xor(x, y)
-  return (x or y) and not (x and y)
-end
-
-local function is_directive(name)
-  return string.sub(name, -1) == '!'
+  return vim.tbl_keys(guard_handlers)
 end
 
 ---@private
+---@param pattern_i integer
+---@param guards ProcessedGuard[]
 ---@param match TSQueryMatch
 ---@param source integer|string
-function Query:match_preds(match, source)
-  local _, pattern = match:info()
-  local preds = self.info.patterns[pattern]
-
-  if not preds then
-    return true
-  end
-
-  local captures = match:captures()
-
-  for _, pred in pairs(preds) do
-    -- Here we only want to return if a predicate DOES NOT match, and
-    -- continue on the other case. This way unknown predicates will not be considered,
-    -- which allows some testing and easier user extensibility (#12173).
-    -- Also, tree-sitter strips the leading # from predicates for us.
-    local is_not = false
-
-    -- Skip over directives... they will get processed after all the predicates.
-    if not is_directive(pred[1]) then
-      local pred_name = pred[1]
-      if pred_name:match('^not%-') then
-        pred_name = pred_name:sub(5)
-        is_not = true
-      end
-
-      local handler = predicate_handlers[pred_name]
-
-      if not handler then
-        error(string.format('No handler for %s', pred[1]))
-        return false
-      end
-
-      local pred_matches = handler(captures, pattern, source, pred)
+function Query:match_guards(pattern_i, guards, captures, source)
```

Comment:
Even if `@private` , still need `_` prefix, especially if it's on a public interface.
```suggestion
function Query:_match_guards(pattern_i, guards, captures, source)
```

Reply:
renamed `apply_directives()` as well.